src/Features/CSharp/Portable/Completion/CompletionProviders/OperatorsAndIndexer/IndexerCompletionProvider.cs

```diff
@@ -0,0 +1,77 @@
+    [ExtensionOrder(After = nameof(ConversionCompletionProvider))]
+    internal class IndexerCompletionProvider : OperatorIndexerCompletionProviderBase
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IndexerCompletionProvider()
+        {
+        }
+
+        protected override int SortingGroupIndex => 1;
+
+        protected override ImmutableArray<CompletionItem> GetCompletionItemsForTypeSymbol(SemanticModel semanticModel,
+            ITypeSymbol container,
+            int position,
+            bool isAccessedByConditionalAccess,
+            CancellationToken cancellationToken)
+        {
+            // We only want to suggest indexer accessible from the cursor position, so we need the containing type at the cursor position,
+            // because the within parameter of GetAccessibleMembersInThisAndBaseTypes() must be an IAssemblySymbol or an INamedTypeSymbol.
+            var containingTypeAtCursorPosition = semanticModel.GetEnclosingSymbol(position, cancellationToken)?.GetContainingTypeOrThis();
+            // We may not be able to identify a containing type, in which case we are conservative and suggest only public indexers
+            var indexers = containingTypeAtCursorPosition is null
+                ? from t in container.GetBaseTypesAndThis()
+                  from i in t.GetIndexers()
+                  where i.HasPublicResultantVisibility()
+                  select i
+                : from p in container.GetAccessibleMembersInThisAndBaseTypes<IPropertySymbol>(containingTypeAtCursorPosition)
+                  where p.IsIndexer
+                  select p;
+            var indexerList = indexers.ToImmutableArray();
+            if (indexerList.Any())
+            {
+                var indexerCompletion = SymbolCompletionItem.CreateWithSymbolId(
+                    displayText: "this",
+                    displayTextSuffix: "[]",
+                    filterText: "this",
+                    sortText: SortText(),
+                    symbols: indexerList,
+                    rules: CompletionItemRules.Default,
+                    contextPosition: position);
+                return ImmutableArray.Create(indexerCompletion);
+            }
+
+            return ImmutableArray<CompletionItem>.Empty;
+        }
+
+        internal override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, TextSpan completionListSpan, char? commitKey, bool disallowAddingImports, CancellationToken cancellationToken)
+        {
+            return
+                await ReplaceDotAndTokenAfterWithTextAsync(document, item, text: "[]", removeConditionalAccess: false, positionOffset: -1, cancellationToken).ConfigureAwait(false) ??
+                await base.GetChangeAsync(document, item, completionListSpan, commitKey, disallowAddingImports, cancellationToken).ConfigureAwait(false);
```

Comment:
The call to `base.GetChangeAsync` is wrong. The three providers act like so:

* Conversion: throw in unexpected code paths
* Indexer, Operator: Fall back to `base.GetChangeAsync`

I think we should unify the providers, but throwing is a bit harsh too. Is there a best practice in case `GetChangeAsync` is (despite all precautions) not able to finish its work?

Reply:
> Is there a best practice in case GetChangeAsync is (despite all precautions) not able to finish its work?

When can it not do it's work?  If it can't do it's can we detect that case when computing the items, and not show them then?

Reply:
In this case (indexer) it is (probably) not an issue. `ReplaceDotAndTokenAfterWithTextAsync` only *fails* if "PreviousTokenIfTouchingWord" is not a dot, which should never be the case.

https://github.com/dotnet/roslyn/blob/e6657b9f9e92f5fcbbbe5c7dd1469cd242481b8b/src/Features/CSharp/Portable/Completion/CompletionProviders/OperatorsAndIndexer/OperatorIndexerCompletionProviderBase.cs#L159-L174

The operator case is more involved but uses the same approach for Infix and Postfix (relying on ReplaceDotAndTokenAfterWithTextAsync) and some custom logic for Prefix:

https://github.com/dotnet/roslyn/blob/e6657b9f9e92f5fcbbbe5c7dd1469cd242481b8b/src/Features/CSharp/Portable/Completion/CompletionProviders/OperatorsAndIndexer/OperatorCompletionProvider.cs#L93-L138

A good solution would be to let ReplaceDotAndTokenAfterWithTextAsync throw, if `GetPreviousTokenIfTouchingWord` is not returning a `dot`.

Reply:
I unified the behavior now. All cases are guarded in `ProvideCompletionsAsync`. So there shouldn't be any surprises in `GetChangeAsync`. Realted commits: 793f1ef3, b894788c and fb964e67.